### PR TITLE
[Added] sending all provisioning events

### LINKED ISSUE #8771: [Server] Send all provisioning events
> @zakisk we're not getting all of Meshery Server's errors sent to the /apis/events, are we?
> 
> For instance, I see this error on stdout, but not via Meshery UI:
> 
> ```
> ERRO[2023-09-13T16:07:12-05:00] the platform local is not supported currently. The supported platforms are: Docker and Kubernetes  app=meshery code=1231 probable-cause="Your platform is not supported for deploying Meshery Adapters" severity=3 short-description="Unable to deploy Meshery Adapter in the current environment" suggested-remediation="Consider using a supported platform for deploying Meshery Adapters"
> ```
> 
> I wonder if you can help with this in a separate PR.
> 
> @MUzairS15 guidance?

@zakisk Refer other parts of server for eg: in design_engine_handler.go or k8s_config_handler.go for sending events

_Originally posted by @MUzairS15 in https://github.com/meshery/meshery/issues/8694#issuecomment-1719607280_

### PR DESCRIPTION
**Pull request by zakisk**
**Notes for Reviewers**

This PR fixes #8771

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by leecalcote** [comment]
@zakisk separately, can you fix this? https://github.com/meshery/meshery/actions/runs/6394536906/job/17356133363?pr=9005

**Comment by leecalcote** [comment]
@MUzairS15 is the extensions.go the right place to wrap events?

**Comment by leecalcote** [comment]
@zakisk I was expecting events to be interlaced throughout other areas of Meshery Server.

**Comment by zakisk** [comment]
@leecalcote 

> @MUzairS15 is the extensions.go the right place to wrap events?

@leecalcote it was just an illustration of how I'm going do it, I will commit more code for other handlers

**Comment by leecalcote** [comment]
> @leecalcote
> 
> > @MUzairS15 is the extensions.go the right place to wrap events?
> 
> @leecalcote it was just an illustration of how I'm going do it, I will commit more code for other handlers

Although, I don't know that this is appropriate for extension events, @zakisk, so I hope that we're not spinning our wheels here, again. 

@theBeginner86 opinions?

**Comment by vishalvivekm** [comment]
@zakisk Let's discuss this on Meshery Dev call. Please add this as an agenda item in the [meeting minutes](https://docs.google.com/document/d/1Dg4Uz9rGTuGHmYZwIt47z1j5x3LmnpxD1M1SgypAkQ0/edit#) if you would. :)

**Review by MUzairS15 — CHANGES_REQUESTED** [review]
@zakisk The issue and the PR description relates to addition of events for provisioning operations but the changes in the PR don’t resonate.

**Comment by MUzairS15** [comment]
As you embark on this task. 
Please keep in mind that there are teo things which needs to be done parallely,
1. Send more events to the UI (wherever required and is missing currently)
2. Update/Add missing fields (short description, long description, probable cause, suggested remediation)
Please do in separate PR/you can add changes for this task starting this PR.

**Comment by zakisk** [comment]
> @zakisk The issue and the PR description relates to addition of events for provisioning operations but the changes in the PR don’t resonate.
@MUzairS15 the code in this PR, at the moment it was a sample that's how I'm gonna do for all handlers. here is a list of API endpoints (provisioning) in those I am gonna add events.
all the sub endpoints of "/api/system/kubernetes/**"
all the sub endpoints of "/api/system/adapter/**"
all the sub endpoints of "/api/telemetry/metrics/grafana/**"
all the sub endpoints of "/api/telemetry/metrics/**"
all the sub endpoints of "/api/prometheus/**"
all the sub endpoints of "/api/telemetry/metrics/**"
all the sub endpoints of "/api/pattern/**"
"/api/schema/resource/{resourceName}"
all the sub endpoints of "/api/meshmodels/**"
all the sub endpoints of "/api/filter/**"
all the sub endpoints of "/api/application/**"
all the sub endpoints of "/api/application/**"
all the sub endpoints of "/api/content/**""
all the sub endpoints of "/api/user/performance/**"

please scrutinize it if I'm missing something or adding inappropriate.

**Comment by zakisk** [comment]
> @zakisk I was expecting events to be interlaced throughout other areas of Meshery Server, but maybe this is where it is missing?

@MUzairS15 the code in this PR, at the moment it was a sample that's how I'm gonna do for all handlers. here is a list of API endpoints (provisioning) in those I am gonna add events.
all the sub endpoints of "/api/system/kubernetes/..."
all the sub endpoints of "/api/system/adapter/..."
all the sub endpoints of "/api/telemetry/metrics/grafana/..."
all the sub endpoints of "/api/telemetry/metrics/..."
all the sub endpoints of "/api/prometheus/..."
all the sub endpoints of "/api/telemetry/metrics/..."
all the sub endpoints of "/api/pattern/..."
"/api/schema/resource/{resourceName}"
all the sub endpoints of "/api/meshmodels/..."
all the sub endpoints of "/api/filter/..."
all the sub endpoints of "/api/application/..."
all the sub endpoints of "/api/application/..."
all the sub endpoints of "/api/content/..."
all the sub endpoints of "/api/user/performance/..."

please scrutinize it if I'm missing something or adding inappropriate.

**Comment by zakisk** [comment]
@leecalcote please see what I've done so far, it is still in progress for other handlers.

**Comment by leecalcote** [comment]
@Philip-21 please be aware...

File: server/handlers/contexts_handler.go [hunk 1]
```diff
@@ -70,15 +95,27 @@ func (h *Handler) GetContext(w http.ResponseWriter, req *http.Request, _ *models
...
+	eventBuilder.ActedUpon(connectionID)
 	h.log.Info("this is being used\n\n\n")
-	val, err := provider.GetK8sContext(token, mux.Vars(req)["id"])
+	val, err := provider.GetK8sContext(token, id)
 	if err != nil {
+		eventBuilder.WithSeverity(events.Error).WithDescription("Encountered an issue while retrieving Kubernetes context.").
+			WithMetadata(map[string]interface{}{
+				"error":                 err,
+				"probable_cause":        "You might have passed incorrect query params or there may be an issue with the connectivity between meshery and kubernetes.",
+				"suggested_remediation": "Please ensure that you've provided valid query params (page, pagesize, search, order), and Meshery can establish connectivity with kubernetes.",
```
Comment (Philip-21) [inline server/handlers/contexts_handler.go:116]: please change this to `suggested remedy`
```suggestion
				"suggested_remedy": "Ensure that you've provided valid query params (page, pagesize, search, order), and Meshery can establish connectivity with kubernetes.",
```

Comment (Philip-21) [reply server/handlers/contexts_handler.go:116]: is there not also another  kind of remedy message , our users may not understand the meaning of 
`provided valid query params (page, pagesize, search, order)`

File: server/handlers/k8sconfig_handler.go [hunk 2]
```diff
@@ -159,16 +158,37 @@ func (h *Handler) deleteK8SConfig(_ *models.User, _ *models.Preference, w http.R
...
-
+	userID := uuid.FromStringOrNil(user.ID)
+	eventBuilder := events.NewEvent().FromUser(userID).FromSystem(*h.SystemID).WithCategory("connection").WithAction("retrieve")
 	k8sConfigBytes, err := readK8sConfigFromBody(req)
 	if err != nil {
+		eventBuilder.WithSeverity(events.Error).WithDescription("Encountered an issue retrieving kubeconfig file from request").
+			WithMetadata(map[string]interface{}{
+				"error":                 err,
+				"probable_cause":        "Request body may contain an empty kubeconfig file.",
+				"suggested_remediation": "Please ensure that you've provided a valid kubeconfig file.",
```
Comment (Philip-21) [inline server/handlers/k8sconfig_handler.go:169]:
```suggestion
				"suggested_remedy": "Ensure you've provided a valid kubeconfig file.",
```

File: server/handlers/k8sconfig_handler.go [hunk 3]
```diff
@@ -238,14 +302,35 @@ func (h *Handler) KubernetesPingHandler(w http.ResponseWriter, req *http.Request
...
+	userID := uuid.FromStringOrNil(user.ID)
+	eventBuilder := events.NewEvent().FromUser(userID).FromSystem(*h.SystemID).WithCategory("connection").WithAction("register")
+
 	k8sConfigBytes, err := readK8sConfigFromBody(req)
 	if err != nil {
+		eventBuilder.WithSeverity(events.Error).WithDescription("Encountered an issue while retrieving kubeconfig file from request").
+			WithMetadata(map[string]interface{}{
+				"error":                 err,
+				"probable_cause":        "Request body may contain empty kubeconfig file.",
+				"suggested_remediation": "Please ensure that you've provided valid kubeconfig file.",
```
Comment (Philip-21) [inline server/handlers/k8sconfig_handler.go:314]:
```suggestion
				"suggested_remediation": "Ensure  you've provided a valid kubeconfig file.",
```

File: server/handlers/mesh_ops_handlers.go [hunk 4]
```diff
@@ -136,36 +138,97 @@ func (h *Handler) MeshAdapterConfigHandler(w http.ResponseWriter, req *http.Requ
...
+		eventBuilder.WithAction("add")
 		meshLocationURL := req.FormValue("meshLocationURL")
 
 		h.log.Debug("meshLocationURL: ", meshLocationURL)
 		if strings.TrimSpace(meshLocationURL) == "" {
+			eventBuilder.WithSeverity(events.Error).WithDescription("Encountered an issue while parsing meshLocationURL").
+				WithMetadata(map[string]interface{}{
+					"error":                 "meshLocationURL is empty",
+					"probable_cause":        "The request body is missing value for `meshLocationURL` field.",
+					"suggested_remediation": "Please ensure that you're providing a valid adapter location URL in the request body",
```
Comment (Philip-21) [inline server/handlers/mesh_ops_handlers.go:154]:
```suggestion
					"suggested_remedy": "Ensure a valid adapter location URL  is provided  in the request body",
```

File: server/handlers/mesh_ops_handlers.go [hunk 5]
```diff
@@ -136,36 +138,97 @@ func (h *Handler) MeshAdapterConfigHandler(w http.ResponseWriter, req *http.Requ
...
 			http.Error(w, ErrAddAdapter.Error(), http.StatusBadRequest)
 			return
 		}
 		meshAdapters, err = h.addAdapter(req.Context(), meshAdapters, prefObj, meshLocationURL, provider)
 		if err != nil {
+			eventBuilder.WithSeverity(events.Error).WithDescription("Encountered an issue while retrieving adapter information.").
+				WithMetadata(map[string]interface{}{
+					"error":                 "Meshery encounters issues when attempting to add and connect with adapter.",
+					"probable_cause":        "The requested adapter service might not be up and running.",
+					"suggested_remediation": "Please ensure that adapter service is up and running and adapter has been deployed successfully.",
```
Comment (Philip-21) [inline server/handlers/mesh_ops_handlers.go:167]:
```suggestion
					"suggested_remediation": "Ensure the specified adapter service is up and running and the adapter has been deployed successfully.",
```

**Comment by zakisk** [comment]
@aabidsofi19 as you're working on Notification in UI, this PR is related to you. please review it. I need your suggestions.

**Review by Philip-21 — CHANGES_REQUESTED** [review]
@zakisk i added some reviews please look into them,  you dont need to add `please`  at the beginning of each statement in the `suggested remedy` aspect, just include the message, it seems more consice. 

Also some messages in the suggested  remedy aspect will some users be able to understand them ?, 
like it may scare them or so.  I am just asking you know . .. What do you think

**Comment by zakisk** [comment]
@Philip-21 thank you, I will take actions accordingly.

File: server/handlers/contexts_handler.go (hunk 1 above)
Comment (zakisk) [reply server/handlers/contexts_handler.go:116]: > is there not also another kind of remedy message , our users may not understand the meaning of `provided valid query params (page, pagesize, search, order)`

Yes, you're right user will be confused about query params :smile:

File: server/handlers/contexts_handler.go [hunk 6]
```diff
@@ -37,6 +41,13 @@ func (h *Handler) GetAllContexts(w http.ResponseWriter, req *http.Request, _ *mo
 	// Don't fetch credentials as UI has no use case.
 	vals, err := provider.GetK8sContexts(token, q.Get("page"), q.Get("pagesize"), q.Get("search"), q.Get("order"), false)
 	if err != nil {
+		eventBuilder.WithSeverity(events.Error).WithDescription("Encountered an issue while retrieving Kubernetes contexts.").
+			WithMetadata(map[string]interface{}{
+				"error":                 err,
+				"probable_cause":        "You might have provided incorrect query params or there may be an issue with connectivity between Meshery and kubernetes.",
+				"suggested_remediation": "Please ensure that you've provided valid query params (page, pagesize, search, order) and Meshery can establish connectivity with kubernetes.",
+			})
+		h.broadcastEvent(eventBuilder, provider, userID)
```
[the remark this answers is not in the document]
Comment (aabidsofi19) [reply server/handlers/contexts_handler.go:50]: @zakisk  can you please confirm if we are using snake_case or PascalCase for metadata properties .  in events response it  mostly PascalCase .

**Comment by aabidsofi19** [comment]
@zakisk  @MUzairS15  @leecalcote  is there a schema defined for the common metadata properties .

**Comment by zakisk** [comment]
> @zakisk @MUzairS15 @leecalcote is there a schema defined for the common metadata properties .

@aabidsofi19 I didn't see anything yet

**Comment by zakisk** [comment]
@aabidsofi19 but It should be, in order to persist consistency, and this is for what I've brought you here.

File: server/handlers/contexts_handler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zakisk) [reply server/handlers/contexts_handler.go:50]: @aabidsofi19 I used the snake case on my own discretion. is it viable for you?

**Comment by Philip-21** [comment]
> @zakisk @MUzairS15 @leecalcote is there a schema defined for the common metadata properties .

@aabidsofi19  No there isnt a schema defined from meshkit its why its an interface to add the additional data for the error message.

**Comment by leecalcote** [comment]
@MUzairS15 up for review.

File: server/handlers/contexts_handler.go (hunk 1 above)
Comment (leecalcote) [reply server/handlers/contexts_handler.go:116]: `suggested_remediation` must remain the same.

The "Please" should be removed.

**Comment by leecalcote** [comment]
Is this using the new Events v2 / event builder?

**Comment by zakisk** [comment]
> @zakisk separately, can you fix this? https://github.com/meshery/meshery/actions/runs/6394536906/job/17356133363?pr=9005

it was an uncertain error, automatically resolved.

**Comment by leecalcote** [comment]
@zakisk will you address this merge conflict?

**Comment by leecalcote** [comment]
@zakisk still in draft?

**Pull request closed without merging**